### DISCUSSION TITLE
Request for Help with GOTM and FABM Configuration

### DISCUSSION THREAD

**Discussion by fdaryabor** [fabm-model/fabm#106] [Q&A] [answered]
Dear All,

I hope this email finds everyone in great spirits! I understand how busy things can get, but I’d be incredibly grateful if someone could help me with an issue I’m facing while configuring gotm.yaml and fabm.yaml.

I’ve attached both files for your reference. My goal is to set up the configuration to calculate DIC and pCO2. Unfortunately, I keep encountering the following error message:

"FATAL ERROR: coupling target is invalid."

I’ve been trying to resolve this on my own, but I haven’t had much luck. If anyone has insights or suggestions, your expertise would be greatly appreciated!

Thank you so much for taking the time to consider my request—it means a lot to me. Please don’t hesitate to reach out if you need any additional details from my side.

Looking forward to your guidance!

Warm regards,
Farshid

Noted: 
[gotm.zip](https://github.com/user-attachments/files/18188789/gotm.zip)
https://github.com/gotm-model/code

fabm.yaml; Section as defined for DIC, pCO2 configuration 
coupling:
      dic: O3/c
  O3:         
    long_name: carbonate
    model: pml/carbonate
    parameters:
      iswCO2: 1                                            # carbonate system diagnostics (0: off, 1: on), default = 1
      iswASFLUX: 1                                         # air-sea CO2 exchange (0: none, 1: Nightingale et al. 2000, 2: Wanninkhof 1992 without chemical enhancement, 3: Wanninkhof 1992 with chemical enhancement, 4: Wanninkhof and McGillis 1999, 5: Wanninkhof 1992 switching to Wanninkhof and McGillis 1999, 6: Wan, default = 6
      iswtalk: 2                                           # alkalinity formulation (1-4: from salinity and temperature, 5: dynamic alkalinity), default = 5
      pHscale: 1                                           # pH scale (1: total, 0: SWS, -1: SWS backward compatible), default = 1
      iswbioalk: 1                                         # use bioalkalinity (0: off, 1: on), default = 1
    initialization:
      c: 2130.0                                            # total dissolved inorganic carbon (mmol C/m^3)
  pCO2_atm:
    model: surface_constant
    parameters:
      standard_name: mole_fraction_of_carbon_dioxide_in_air
      value: 400

gotm.yaml; fabm setting

fabm:
   use: true
   freshwater_impact: true
   feedbacks:
      shade: false
      albedo: true
      surface_drag: true
   repair_state: true
   configuration_method: yaml
   input:
     msi_ergom1/po:
       method: file                
       constant_value: 0.4         
       file: phosphate.dat  
       relax_tau: 3.15000000E+07          
       column: 1                   
     msi_ergom1/nn:
       method: file                
       constant_value: 8.         
       file: nitrate.dat 
       relax_tau: 3.15000000E+07          
       column: 1                   
     msi_ergom1/aa: 
       method: constant
       constant_value: 0.001
       file:  
       column: 4 
     msi_ergom1/o2:
       method: file
       constant_value: 300.
       file: oxygen.dat
       relax_tau: 3.15000000E+07
       column: 1
     O3/c:
       method: constant
       constant_value: 2130.0
       file:
       column: 1
     mole_fraction_of_carbon_dioxide_in_air:
       method: constant
       constant_value: 385.0
       file:
       column: 1

**Comment by jornbr** [comment]
Hi Farshid,

Is there any earlier eror message printed before the "FATAL ERROR"? I do not see anything strange in the yaml snippets you have included above, so can you attach your complete `fabm.yaml` file? And what version of FABM are you using?

Best,

Jorn

**Comment by fdaryabor** [comment]
Hi Jorn,

Good to hear from you.

The error has popped up again, and it is as follows:
$~/source/repos/GOTM/setup$ cat gotm.log
STOP 1
 ------------------------------------------------------------------------
 GOTM started on 2024/12/19 at 11:19:27
 ------------------------------------------------------------------------
    initialize_gotm
 ------------------------------------------------------------------------
        Reading configuration from: gotm.yaml
        configuring modules ....
    init_airsea_yaml
        done
    init_observations_yaml
    init_stokes_drift_yaml
        done
    init_turbulence_yaml
        done.
    init_gotm_fabm_yaml
        Reading configuration from:
            fabm.yaml
        done.
    init_meanflow_yaml
        done
    init_eqstate_yaml
    init_density()
        Linearized - TEOS10 - using S0, T0, p0
            S0=       35.000000000000000
            T0=       10.000000000000000
            p0=       0.0000000000000000
            rho0=     1027.0000000000000
            alpha=    1.6625612540220982E-004
            beta=     7.5366784499087118E-004
            cp=       3991.8679571196299
        done.
 FATAL ERROR: model /msi_ergom1, process_coupling_tasks: Coupling target
"O3/c" for "dic" was not found.
 Initializing msi_ergom1...
    model type: msi_ergom1
    initialization succeeded.
 Initializing O3...
    model type: pml/carbonate
    initialization succeeded.
 Initializing pCO2_atm...
    model type: surface_constant
    initialization succeeded.


The code I use was downloaded and compiled from the below repository;
https://github.com/gotm-model/code,

Please find attached 'fabm.yaml' and 'gotm.yaml'

Thanks Jorn for your support,

Cheers,

Farshid

**Comment by fdaryabor** [comment]
Dear Jorn, if you can find the attached files I can past below;

'fabm.yaml'

$~/source/repos/GOTM/setup$ cat fabm.yaml
instances:
  msi_ergom1:
    parameters:
      env_type: marine          # (Define environment type, either fresh or marine), default = marine
      calc_dic: true            # (Decide whether to calculate DIC), default = false
      dic_variable: dic         # (Define DIC variable)
      nitrogen_fixation: true   # (Nitrogen fixation of cyanos), default = true
      wdz: -4.5                 # Detritus sinking velocity (m/d), default = -4.5
      wpz: -0.5                 # Diatoms sinking velocity (m/d), default = -0.5
      wfz: 0.0                  # Zooplankton sinking velocity (m/d), default = 0.0
      wbz: 0.1                  # Cyanobacteria sinking velocity (positive) (m/d), default = 0.1
      wpo4: -1.0                # P-Fe in water sinking velocity (m/d), default = -1.0
      sfl_po: 0.0015            # Constant surface phosphate flux (mmol p/m2/d), default = 0.0015
      sfl_nn: 0.083             # Constant surface nitrate flux (mmol n/m2/d), default = 0.083
      sfl_aa: 0.06              # Constant surface ammonium flux (mmol n/m2/d), default = 0.06
      nb: 0.01                  # Phytoplankton excretion rate (pl -> aa) (1/d), default = 0.01
      deltao: 0.02              # Phytoplankton mortality rate (pl -> dd) (1/d), default = 0.02
      nue: 0.01                 # Zooplankton respiration rate (zz -> aa) (m3/d/mmol n), default = 0.01
      sigma_b: 0.03             # Zooplankton mortality rate (zz -> dd) (m3/d/mmol n), default = 0.03
      dn: 0.003                 # Detritus mineralization rate (dd -> aa) (1/d), default = 0.003
      dn_sed: 0.002             # Sediment mineralization rate (fl -> aa) (1/d), default = 0.002
      rp0: 1.3                  # Diatoms uptake rate (1/d), default = 1.3
      rf0: 0.4                  # Flagellates uptake rate (1/d), default = 0.4
      rb0: 0.75                 # Cyanobacteria uptake rate (1/d), default = 0.75
      cyanotll: 13.5            # Cyanobacteria lower temperature limit (deg C), default = 13.5
      cyanosll: 1.0             # Cyanobacteria lower salinity limit (PSU), default = 1.0
      cyanosul: 10.0            # Cyanobacteria upper salinity limit (PSU), default = 10.0
      flagtll: 100.0            # Flagellates half-saturation temp, squared ((deg C)²), default = 100.0
      alphap: 0.25              # Half-saturation const, diatoms (mmol n/m3), default = 0.25
      alphaf: 0.1               # Half-saturation const, flagellates (mmol n/m3), default = 0.1
      alphab: 0.4               # Half-saturation const, cyanobacteria (mmol n/m3), default = 0.4
      Yc_diat: 6.25             # Carbon to Chla ratio (umol C/mg Chla), diatoms, default = 6.25
      Yc_flag: 6.25             # Carbon to Chla ratio (umol C/mg Chla), flagellates, default = 6.25
      Yc_cyan: 6.25             # Carbon to Chla ratio (umol C/mg Chla), cyanobacteria, default = 6.25
      iv: 1.2                   # Ivlev constant, quadratic (1/(mmol n/m3)³), default = 1.2
      graz: 0.5                 # Zooplankton grazing rate (1/d), default = 0.5
      toptz: 20.0               # Optimal temperature for grazing (deg C), default = 20.0
      zcl1: 50.0                # Zooplankton closure parameter (-), default = 50.0
      p0: 0.001                 # Diatoms background value (mmol n/m3), default = 0.001
      f0: 0.001                 # Flagellates background value (mmol n/m3), default = 0.001
      b0: 0.001                 # Cyanobacteria background value (mmol n/m3), default = 0.001
      z0: 0.001                 # Zooplankton background value (mmol n/m3), default = 0.001
      imin_di: 35.0             # Minimal optimal light radiation, diatoms (W/m²), default = 35.0
      imin: 50.0                # Minimal optimal light radiation, others (W/m²), default = 50.0
      kc: 0.03                  # Attenuation constant for self-shading effect (m²/mmol n)
      q10_rec: 0.15             # Sediment recycling Q10 rule factor (-), default = 0.15
      ade_r0: 0.1               # Chemoautolithotrophic denitrification rate (1/d), default = 0.1
      alphaade: 1.0             # Half-saturation constant for ade (mmol n/m³), default = 1.0
      q10_recs: 0.175           # Sediment recycling Q10 rule factor (-), default = 0.175
      rfr: 0.0625               # Redfield ratio P/N (-), default = 0.0625
      rfc: 6.625                # Redfield ratio C/N (-), default = 6.625
      sedrate: 2.25             # Detritus sedimentation rate (m/d), default = 2.25
      erorate: 6.0              # Sediment erosion rate (m/d), default = 6.0
      sedratepo4: 0.5           # P-Fe sedimentation rate (m/d), default = 0.5
      eroratepo4: 6.0           # P-Fe erosion rate (m/d), default = 6.0
      po4ret: 0.18              # Phosphate retention rate, oxic sediments (-), default = 0.18
      pburialrate: 0.007        # Phosphate burial rate (-), default = 0.007
      fl_burialrate: 0.001      # Sediment burial rate (-), default = 0.001
      pliberationrate: 0.1      # Phosphate liberation rate, anoxic sediments (-), default = 0.1
      ipo4th: 100.0             # Increased phosphate burial threshold (mmol p/m²), default = 100.0
      maxsed: 1000.0            # Maximum active sediment concentration (-), default = 1000.0
      br0: 0.03                 # Bioresuspension rate (1/d), default = 0.03
      fds: 0.7                  # Sediment denitrification rate, oxic (-), default = 0.7
      pvel: 5.0                 # Piston velocity (m/d), default = 5.0
      newflux: 2                # Oxygen flux type (-), default = 2
      tau_crit: 0.07            # Critical shear stress (N/m²), default = 0.07
    initialization:
      pp: 0.001                 # Diatoms (mmol n/m³)
      ff: 0.001                 # Flagellates (mmol n/m³)
      bb: 0.001                 # Cyanobacteria (mmol n/m³)
      zz: 0.001                 # Zooplankton (mmol n/m³)
      dd: 0.01                  # Detritus (mmol n/m³)
      aa: 0.001                 # Ammonium (mmol n/m³)
      nn: 8.0                   # Nitrate (mmol n/m³)
      po: 1.0                   # Phosphate (mmol p/m³)
      o2: 300.0                 # Oxygen (mmol o₂/m³)
      fl: 100.0                 # Fluff (mmol n/m²)
      pb: 0.001                 # PFe_s (mmol p/m²)
      pw: 0.001                 # PFe_w (mmol p/m³)
    coupling:
      dic: O3/c
  O3:
    long_name: carbonate
    model: pml/carbonate
    parameters:
      iswCO2: 1                 # carbonate system diagnostics (0: off, 1: on), default = 1
      iswASFLUX: 1              # air-sea CO2 exchange(0: none, 1:, 2: without chemical enhancement, 3: with chemical enhancement)
      iswtalk: 2                # alkalinity formulation (1-4: from salinity and temperature, 5: dynamic alkalinity), default = 5
      pHscale: 1                # pH scale (1: total, 0: SWS, -1: SWS backward compatible), default = 1
      iswbioalk: 1              # use bioalkalinity (0: off, 1: on), default = 1
    initialization:
      c: 2130.0                 # total dissolved inorganic carbon (mmol C/m^3)
  pCO2_atm:
    model: surface_constant
    parameters:
      standard_name: mole_fraction_of_carbon_dioxide_in_air
      value: 400

--------------------------------------------------------------------------------------------------------------------------------------------------------

'gotm.yaml'

$~/source/repos/GOTM/setup$ cat gotm.yaml
version: 7
title: GOTM-ERGOM at Arkona
location:
  name: Arikona station
  latitude: 54.883
  longitude: 13.867
  depth: 45.0
time:
  start: 2019-01-01 00:00:00
  stop: 2019-04-01 00:00:00
  dt: 900.0
grid:
  nlev: 71
  method: file_sigma
  ddu: 1.0
  ddl: 1.0
  file: grid.dat
temperature:
  method: file
  constant_value: 15.0
  file: tprof.dat
  column: 1
  two_layer:
    z_s: 30.0
    t_s: 15.0
    z_b: 40.0
    t_b: 10.0
  NN: 0.000256
  relax:
    tau: 2592000.0
    tau_s: 2592000.0
    tau_b: 2592000.0
salinity:
  method: file
  constant_value: 35.0
  file: sprof.dat
  column: 1
  two_layer:
    z_s: 30.0
    s_s: 35.0
    z_b: 40.0
    s_b: 36.0
  NN: 0.000256
  relax:
    tau: 2592000.0
    tau_s: 25292000.0
    tau_b: 25292000.0
surface:
  fluxes:
    method: fairall
    heat:
      method: constant
      constant_value: 0.0
      file:
      column: 1
    tx:
      method: constant
      constant_value: 0.0
      file:
      column: 1
    ty:
      method: constant
      constant_value: 0.0
      file:
      column: 1
  u10:
    method: file
    constant_value: 0.0
    file: meteo.dat
    column: 1
  v10:
    method: file
    constant_value: 0.0
    file: meteo.dat
    column: 2
  ssuv_method: absolute
  airp:
    method: file
    constant_value: 0.0
    file: meteo.dat
    column: 3
    scale_factor: 100.0
  airt:
    method: file
    constant_value: 0.0
    file: meteo.dat
    column: 4
  hum:
    method: file
    constant_value: 0.0
    file: meteo.dat
    column: 5
    type: dew_point
  cloud:
    method: file
    constant_value: 0.0
    file: meteo.dat
    column: 6
  precip:
    method: file
    constant_value: 0.0
    file: precip.dat
    column: 1
    flux_impact: true
  calc_evaporation: true
  swr:
    method: file
    constant_value: 0.0
    file: ssr.dat
    column: 1
  longwave_radiation:
    method: clark
    file:
    column: 1
  albedo:
    method: constant
    constant_value: 0.0
  roughness:
    charnock: false
    charnock_val: 1400.0
    z0s_min: 0.02
bottom:
  h0b: 0.05
light_extinction:
  method: jerlov-iii
  A:
    method: constant
    constant_value: 0.7
    file:
    column: 1
  g1:
    method: constant
    constant_value: 0.4
    file:
    column: 1
  g2:
    method: constant
    constant_value: 8.0
    file:
    column: 1
turbulence:
  turb_method: second_order
  tke_method: tke
  len_scale_method: dissipation
  stab_method: schumann_gerz
  turb_param:
    cm0_fix: 0.5477
    Prandtl0_fix: 0.74
    cw: 100.0
    compute_kappa: false
    kappa: 0.4
    compute_c3: true
    Ri_st: 0.25
    length_lim: true
    galp: 0.53
    const_num: 0.0005
    const_nuh: 0.0005
    k_min: 1.0e-06
    eps_min: 1.0e-12
    kb_min: 1.0e-10
    epsb_min: 1.0e-14
  generic:
    gen_m: 1.0
    gen_n: -0.67
    cpsi1: 1.0
    cpsi2: 1.22
    cpsi3minus: 0.05
    sig_kpsi: 0.8
    sig_psi: 1.07
  keps:
    ce3minus: -0.4
  my:
    length: linear
  scnd:
    method: quasi_eq
  iw:
    alpha: 0.7
mimic_3d:
  ext_pressure:
    type: velocity
    dpdx:
      method: file
      constant_value: 0.0
      file: ext_press.dat
      column: 2
      tidal:
        amp_1: 0.2429
        phase_1: 32171.1
        amp_2: 0.085
        phase_2: 36650.0
    dpdy:
      method: file
      constant_value: 0.0
      file: ext_press.dat
      column: 3
      tidal:
        amp_1: 0.0734
        phase_1: 14490.5
        amp_2: 0.0257
        phase_2: 19567.8
    h:
      method: file
      constant_value: 1.0
      file: ext_press.dat
      column: 1
    period_1: 44714.0
    period_2: 43200.0
  zeta:
    method: file
    constant_value: 0.0
    file: zeta.dat
    column: 1
    tidal:
      period_1: 44714.0
      amp_1: 1.4287
      phase_1: 28645.5
      period_2: 43200.0
      amp_2: 0.5
      phase_2: 33243.4
  w:
    max:
      method: off
      constant_value: 0.0
      file:
      column: 1
    height:
      method: constant
      constant_value: 0.0
      file:
      column: 1
    adv_discr: superbee
  int_pressure:
    t_adv: false
    s_adv: false
    gradients:
      dtdx:
        method: off
        constant_value: 0.0
        file:
        column: 1
      dtdy:
        method: off
        constant_value: 0.0
        file:
        column: 1
      dsdx:
        method: off
        constant_value: 0.0
        file:
        column: 1
      dsdy:
        method: off
        constant_value: 0.0
        file:
        column: 1
o2:
  scale_factor: 31.25
fabm:
   use: true
   freshwater_impact: true
   feedbacks:
      shade: false
      albedo: true
      surface_drag: true
   repair_state: true
   configuration_method: yaml
   input:
     msi_ergom1/po:
       method: file
       constant_value: 0.4
       file: phosphate.dat
       relax_tau: 3.15000000E+07
       column: 1
     msi_ergom1/nn:
       method: file
       constant_value: 8.
       file: nitrate.dat
       relax_tau: 3.15000000E+07
       column: 1
     msi_ergom1/aa:
       method: constant
       constant_value: 0.001
       file:
       column: 4
     msi_ergom1/o2:
       method: file
       constant_value: 300.
       file: oxygen.dat
       relax_tau: 3.15000000E+07
       column: 1
     O3/c:
       method: constant
       constant_value: 2130.0
       file:
       column: 1
     mole_fraction_of_carbon_dioxide_in_air:
       method: constant
       constant_value: 385.0
       file:
       column: 1
equation_of_state:
  method: linear_teos-10    # or linear_custom
  rho0: 1027.0
  linear:
    T0: 10.0
    S0: 35.0
    p0: 0.0
    alpha: -0.17            # thermal expansion coefficient (Replace dtr0 with alpha)
    beta: 0.78              # saline expansion coefficient (Replace dsr0 with beta)
restart:
  load: false
output:
  result:
    format: netcdf
    time_unit: day
    time_step: 1
    time_method: mean
    variables:
      - source: /*

**Comment by jornbr** [comment, marked as answer]
Hi Farshid,

The issue is due to having a mixture of the old and new PML carbonate model in your `fabm.yaml`. If you replace `model: pml/carbonate` (the old model version) with `model: ersem/carbonate` (the new version), and you have compiled FABM with [ERSEM](https://github.com/pmlmodelling/ersem) included, it all works fine.

Alternatively, if you really do want to use the old carbonate model, you'd have to change the `O3` section to something like:

```
  O3:
    long_name: carbonate
    model: pml/carbonate
    initialization:
      dic: 2130.0 # total dissolved inorganic carbon (mmol C/m^3)
```

And your coupling section of `msi_ergom1` to:

```
    coupling:
      dic: O3/dic
```

However, I note the settings in your `gotm.yaml` file (e.g., the mention of `O3/c`) are also specific to the new carbonate model, so likely that's the one you want.

Best,

Jorn

**Comment by fdaryabor** [comment]
Hi Jorn,

Thank you very much for your helpful suggestion.

After recompiling GOTM-FABM with [ERSEM](https://github.com/pmlmodelling/ersem) included, the model ran exactly as expected, similar to my previous runs. Unfortunately, I had forgotten to compile ERSEM with GOTM earlier, so I had been using the carbonate system configuration, as shown below:

coupling:
  dic: O3/c
O3:
  long_name: carbonate
  model: ersem/carbonate
  parameters:
    iswCO2: 1                                             # carbonate system diagnostics (0: off, 1: on), default = 1
    iswASFLUX: 1                                          # air-sea CO2 exchange (0: none, 1: Nightingale et al. 2000, 2: Wanninkhof 1992 without chemical enhancement, 3: Wanninkhof 1992 with chemical enhancement, 4: Wanninkhof and McGillis 1999, 5: Wanninkhof 1992 switching to Wanninkhof and McGillis 1999, 6: Wan, default = 6)
    iswtalk: 2                                            # alkalinity formulation (1-4: from salinity and temperature, 5: dynamic alkalinity), default = 5
    pHscale: 1                                            # pH scale (1: total, 0: SWS, -1: SWS backward compatible), default = 1
    iswbioalk: 1                                          # use bioalkalinity (0: off, 1: on), default = 1
  initialization:
    c: 2130.0                                              # total dissolved inorganic carbon (mmol C/m^3)

Below, I leave the command used for recompiling:

cmake ../code/ -DFABM_INSTITUTES="msi;ersem;pml;au;bb;gotm;su" -DFABM_ERSEM_BASE=../../ersem/

make install -j4

To make sure ersem/carbonate has already compiled 'find -name "*ersem*" -or -name "*carbonate*"'








CHEERS,
